File: Lista_5/lista5_ex9.c
/* Exercício 9:
Uma matriz quadrada de inteiros é um quadrado mágico se a soma dos elementos de cada linha,
a soma dos elementos de cada coluna, a soma dos elementos da diagonal principal e da diagonal secundária são todos iguais. 
A matriz abaixo é um exemplo de quadrado mágico:

3  4 8
10 5 0
2  6 7

Escreva uma função que recebe como parâmetro uma matriz quadrada de tamanho no máximo 30 × 30, 
e suas dimensões n, e determina se ela é um quadrado mágico devolvendo 1 caso seja e 0 caso contrário. 
O protótipo da função deve ser:
int magico(int mat[30][30], int n);
*/
#include <stdio.h>

int magico(int mat[30][30], int n);

int main() {
int mat[30][30], n, i, j, r;

printf("Qual o valor do n, para a dimensao n x n da matriz: ");
scanf("%d", &n);

for(i = 0; i < n; i++) {
    for(j = 0; j < n; j++) {
        printf("\nInsira elemento na posicao (linha: %d, coluna: %d): ", i, j);
        scanf("%d", &mat[i][j]);
    }
}

printf("\nMatriz criada:\n");
for(i = 0; i < n; i++) {
    printf("| ");
    for(j = 0; j < n; j++) {
        printf("%d ", mat[i][j]);
    }
    printf("|\n");
}

r = magico(mat, n);

if(r == 1) {
    printf("\nA matriz e um quadrado magico!\n");
} else {
    printf("\nA matriz nao e um quadrado magico!\n");
}

return 0;
}

int magico(int mat[30][30], int n) {
    int temp[62], i, j, sum=0, c=0;
    ///Soma linhas;
    for(i = 0; i < n; i++) {
        for(j = 0; j < n; j++) {
            sum = sum + mat[i][j];
        }
        temp[c] = sum;
        c++;
        sum=0;
    }
    ///Soma colunas;
    for(i = 0; i < n; i++) {
        for(j = 0; j < n; j++) {
            sum = sum + mat[j][i];
        }
        temp[c] = sum;
        c++;
        sum=0;
    }
    ///Soma diagonal principal;
    for(i = 0; i < n; i++) {
        sum = sum + mat[i][i];
    }
    temp[c] = sum;
    c++;
    sum=0;
    ///Soma diagonal secundaria;
    for(i = 0, j = n-1; i < n, j >= 0; i++, j--) {
        sum = sum + mat[i][i];
    }
    temp[c] = sum;
    ///Comparacao final
    for(i = 1; i < c; i++) {
        if(temp[0] != temp[i]) {
            return 0;
        }
    }
    return 1;
}

File: Lista_5/Lista5_ex9.py
def cria_matriz(n):
	matriz = []
	
	for i in range(n):
		linha = []
		for j in range(n):
			numero = int(input("Linha {} e Coluna {}: ".format(i+1, j+1)))
			linha.append(numero)
		matriz.append(linha)
	return matriz

def mostra_matriz(matriz, n):
	for i in range(n):
		for j in range(n):
			print(matriz[i][j], end=" ")
		print()

def magico(matriz, n):
	vetor = []
	soma = 0
	#Soma linhas
	for i in range(n):
		for j in range(n):
			soma = soma + matriz[i][j] 
		vetor.append(soma)
		soma = 0
	#Soma colunas
	for i in range(n):
		for j in range(n):
			soma = soma + matriz[j][i] 
		vetor.append(soma)
		soma = 0
	#Soma diagonal principal
	for i in range(n):
		soma = soma + matriz[i][i] 
	vetor.append(soma)
	soma = 0
	#Soma diagonal secundaria
	i = 0
	j = n - 1
	while i < n and j >= 0:
		soma = soma + matriz[i][j]
		i += 1
		j -= 1
	vetor.append(soma)
	#Comparacao final
	for i in range(len(vetor)):
		if vetor[0] != vetor[i]:
			return False
	return True

def main():
	matriz = []
	n = int(input("Qual o valor do n, para a dimensao n x n da matriz: "))

	print("\nInsira os valores na matriz:")
	matriz = cria_matriz(n)

	print("\nMatriz:")
	mostra_matriz(matriz, n)

	verifica = magico(matriz, n)

	if verifica:
		print("\nA matriz e um quadrado magico!\n")
	else:
		print("\nA matriz nao e um quadrado magico!\n") 

if __name__ == "__main__":
	main()
